libfwupdplugin/fu-udev-device.c

```diff
@@ -459,71 +455,37 @@ fu_udev_device_probe(FuDevice *device, GError **error)
 	}
 
 	/* add GUIDs in order of priority */
-	if (priv->vendor != 0x0000 && priv->model != 0x0000 && priv->subsystem_vendor != 0x0000 &&
-	    priv->subsystem_model != 0x0000) {
-		g_autofree gchar *devid1 = NULL;
-		g_autofree gchar *devid2 = NULL;
-		devid1 = g_strdup_printf("%s\\VEN_%04X&DEV_%04X&SUBSYS_%04X%04X&REV_%02X",
-					 subsystem,
-					 priv->vendor,
-					 priv->model,
-					 priv->subsystem_vendor,
-					 priv->subsystem_model,
-					 priv->revision);
-		fu_device_add_instance_id(device, devid1);
-		devid2 = g_strdup_printf("%s\\VEN_%04X&DEV_%04X&SUBSYS_%04X%04X",
-					 subsystem,
-					 priv->vendor,
-					 priv->model,
-					 priv->subsystem_vendor,
-					 priv->subsystem_model);
-		fu_device_add_instance_id(device, devid2);
-	}
-	if (priv->vendor != 0x0000 && priv->model != 0x0000) {
-		g_autofree gchar *devid = NULL;
-		devid = g_strdup_printf("%s\\VEN_%04X&DEV_%04X&REV_%02X",
-					subsystem,
-					priv->vendor,
-					priv->model,
-					priv->revision);
-		fu_device_add_instance_id(device, devid);
-	}
-	if (priv->vendor != 0x0000 && priv->model != 0x0000) {
-		g_autofree gchar *devid = NULL;
-		devid =
-		    g_strdup_printf("%s\\VEN_%04X&DEV_%04X", subsystem, priv->vendor, priv->model);
-		fu_device_add_instance_id(device, devid);
-	}
-	if (priv->vendor != 0x0000) {
-		g_autofree gchar *devid = NULL;
-		devid = g_strdup_printf("%s\\VEN_%04X", subsystem, priv->vendor);
-		fu_device_add_instance_id_full(device, devid, FU_DEVICE_INSTANCE_FLAG_ONLY_QUIRKS);
-	}
+	if (priv->vendor != 0x0000)
+		fu_device_instance_assign_uint16(device, "VEN", priv->vendor);
+	if (priv->model != 0x0000)
+		fu_device_instance_assign_uint16(device, "DEV", priv->model);
+	if (priv->subsystem_vendor != 0x0000 && priv->subsystem_model != 0x0000) {
+		g_autofree gchar *subsys =
+		    g_strdup_printf("%04X%04X", priv->subsystem_vendor, priv->subsystem_model);
+		fu_device_instance_assign(device, "SUBSYS", subsys);
+	}
+	fu_device_instance_assign_uint8(device, "REV", priv->revision);
+
+	fu_device_instance_build_quirk(device, subsystem, "VEN", NULL);
+	fu_device_instance_build(device, subsystem, "VEN", "DEV", NULL);
+	fu_device_instance_build(device, subsystem, "VEN", "DEV", "REV", NULL);
+	fu_device_instance_build(device, subsystem, "VEN", "DEV", "SUBSYS", NULL);
+	fu_device_instance_build(device, subsystem, "VEN", "DEV", "SUBSYS", "REV", NULL);
```

Comment:
there's something subtly different here - the old code wouldn't make any IDs if you didn't have a vendor.  This code will now try to make an ID even with vendor not set.

Reply:
Nope, if `priv->vendor == 0x0000` then we don't call `fu_device_instance_assign_uint16(device, "VEN", priv->vendor);` and if the key isn't set then we return from `fu_device_instance_build` without adding anything. Right?

Reply:
Oh I guess that's true, but maybe the thought process that lead me here means that `fu_device_instance_build` should not be `void`.  If no strings are going to be made, I would expect some sort of failure that the caller would know about rather than silently doing nothing.

Reply:
Returning just True/False, or setting an optional GError too? I guess it makes sense, as some plugins *need* specific IDs to make sense.

Reply:
GError too make, it will make our lives easier in the future maybe

Reply:
Agree, this works well. Fixed.